### PR TITLE
Fixes #2068 remove indexes to fix

### LINKED ISSUE #2068: Partial Preload trying to access incorrect shop page urls (WooCommerce)
When you publish a product in WooCommerce or update it, `Partial_Preload` is trying to access some incorrect URLs for the page set as the Shop page in WooCommerce. 

On every product publish/update, these 3 URLs 404: 
404s

Steps to reproduce: 

- Activate WooCommerce
- Go to WooCommerce > Settings > Products: make sure the Shop page has been set (it can be set to any page): https://jmp.sh/8midTHf
- Make sure preload is active in WP Rocket
- Update a product
- Use whichever tool you like to log 404 errors (I used Redirection plugin)
- Observe that the URLS indicated give a 404: https://jmp.sh/dGHBk7w

Tickets:
https://secure.helpscout.net/conversation/986778716/127894?folderId=377611
https://secure.helpscout.net/conversation/983825572/127434?folderId=377611

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Fix a problem when updating a product where the product page is not reload in the preload.

For that we did create a filter and a callback that remove index.html and index.html_gzip from  urls added to the preload.

Fixes #2068

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Automated Test
- [ ] Manual Test

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Preload/Database/Queries/Cache.php [hunk 1]
```diff
@@ -124,6 +124,14 @@ public function __construct( Logger $logger, $query = [] ) {
...
 		$url = untrailingslashit( strtok( $resource['url'], '?' ) );
 
+		/**
+		 * Format the url.
+		 *
+		 * @param string $url url to format.
+		 * @return string
+		 */
+		$url = apply_filters( 'rocket_preload_format_url', $url );
+
```
Comment (jeawhanlee) [inline inc/Engine/Preload/Database/Queries/Cache.php]: `$url` will return an empty string when it matches an index if I'm right, should we not also add a check here to bail out when `$url` is an empty string?

File: inc/Engine/Preload/Database/Queries/Cache.php [hunk 2]
```diff
@@ -184,6 +192,14 @@ public function create_or_update( array $resource ) {
...
 		$url = strtok( $resource['url'], '?' );
 
+		/**
+		 * Format the url.
+		 *
+		 * @param string $url url to format.
+		 * @return string
+		 */
+		$url = apply_filters( 'rocket_preload_format_url', $url );
+
```
Comment (jeawhanlee) [inline inc/Engine/Preload/Database/Queries/Cache.php:214]: here too

File: inc/Engine/Preload/Database/Queries/Cache.php (hunk 1 above)
Comment (CrochetFeve0251) [reply inc/Engine/Preload/Database/Queries/Cache.php]: @jeawhanlee we should have only full link with domains so we will never return an empty string

File: inc/Engine/Preload/Subscriber.php [hunk 3]
```diff
@@ -375,4 +376,15 @@ public function preload_after_automatic_cache_purge( $deleted ) {
...
+
+	/**
+	 * Remove index from url.
+	 *
+	 * @param string $url url to reformat.
+	 *
+	 * @return string
+	 */
+	public function format_preload_url( string $url ) {
+		return preg_replace( '/(index\.html$)|(index\.html_gzip$)/', '', $url );
```
Comment (remyperona) [inline inc/Engine/Preload/Subscriber.php]: This also needs to take into account the following:
- index-https.html(_gzip)

**Review by remyperona — CHANGES_REQUESTED** [review]
We should also fix an incorrect value in `inc/common/purge.php`, `rocket_clean_post()`, line 74, the following line:
`$purge_urls[]      = $post_type_archive . $filename . $GLOBALS['wp_rewrite']->pagination_base;`

should be:
`$purge_urls[]      = $post_type_archive . $GLOBALS['wp_rewrite']->pagination_base;`

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 @piotrbak 
Currently, on the branch no index URLs but the following is happening (independent on permalinks with/without `/ `)
1- Preload is enabled => (both shop and product exist in cache table without/at the end) https://newer.rocketlabsqa.ovh/shop-2 and https://newer.rocketlabsqa.ovh/product/spanish-session
2- edit the product
3- 2 new URLs added for the shop 
`https://newer.rocketlabsqa.ovh/shop-2/` and `https://newer.rocketlabsqa.ovh/shop-2/page` while the shop page in 1 is still completed => after some time `/shop-2/ `will be stuck in progress while `/shop-2/page` will be removed from the table
4- once `https://newer.rocketlabsqa.ovh/product/spanish-session` is completed, the original `https://newer.rocketlabsqa.ovh/shop-2 ` date will be updated

Didn't expect those URLs in step 3, what do you think?

**Comment by piotrbak** [comment]
@Mai-Saad For 3. The /shop-2/ is stuck `in-progress`? If the page is accessible, it should not happen at all.

When it comes to the `/shop-2/page/`, it doesn't exist, right?

@CrochetFeve0251 Does the `/shop-2/page/` come from this place?
https://github.com/wp-media/wp-rocket/blob/c53d551c38788c3ac1e171cf60b52af4c777c204/inc/common/purge.php#L74

**Comment by CrochetFeve0251** [comment]
> @Mai-Saad For 3. The /shop-2/ is stuck `in-progress`? If the page is accessible, it should not happen at all.
> 
> When it comes to the `/shop-2/page/`, it doesn't exist, right?
> 
> @CrochetFeve0251 Does the `/shop-2/page/` come from this place?
> 
> https://github.com/wp-media/wp-rocket/blob/c53d551c38788c3ac1e171cf60b52af4c777c204/inc/common/purge.php#L74

@Mai-Saad @piotrbak there was an issue on the formating from the URL added in the database due to thev order functions where called. I fixed that and you can try now it is working fine.

**Comment by Mai-Saad** [comment]
@piotrbak @CrochetFeve0251 

> When it comes to the /shop-2/page/, it doesn't exist, right?

It is still added when edit/add product as pending, then will be removed after being in-progress for awhile. any idea why is it added?

**Comment by CrochetFeve0251** [comment]
> @piotrbak @CrochetFeve0251
> 
> > When it comes to the /shop-2/page/, it doesn't exist, right?
> 
> It is still added when edit/add product as pending, then will be removed after being in-progress for awhile. any idea why is it added?

@Mai-Saad We remove the index in the url but we didn't deleted that part.

**Comment by piotrbak** [comment]
@CrochetFeve0251 Would it be possible to remove `/page/` without any pagination number (`/page/2/`, etc.) from the archive page there?

**Comment by piotrbak** [comment]
Related Slack convo:
https://wp-media.slack.com/archives/GUT7FLHF1/p1667910845418149

**Comment by piotrbak** [comment]
Since for now we don't have easy workaround to solve problem mentioned by @Mai-Saad, let's move forward with the current solution as it's a major improvement anyways.

**Comment by Mai-Saad** [comment]
Working fine but if We updated WPR while having the `/shop/index.html` in the database, once those links are processed, we will have a Warning in debug.log, is there any way to prevent this? @CrochetFeve0251 
```
[24-Nov-2022 10:39:50 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/newer.rocketlabsqa.ovh/shop-2/index-https.html/index-https.html_gzip) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 96
[24-Nov-2022 10:39:50 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/newer.rocketlabsqa.ovh/shop-2/index-https.html/index-https.html) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 100
```

**Comment by CrochetFeve0251** [comment]
> Working fine but if We updated WPR while having the `/shop/index.html` in the database, once those links are processed, we will have a Warning in debug.log, is there any way to prevent this? @CrochetFeve0251
> 
> ```
> [24-Nov-2022 10:39:50 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/newer.rocketlabsqa.ovh/shop-2/index-https.html/index-https.html_gzip) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 96
> [24-Nov-2022 10:39:50 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/newer.rocketlabsqa.ovh/shop-2/index-https.html/index-https.html) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/newer.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 100
> ```

@Mai-Saad @vmanthos and I can't reproduce that warning on a different server than the QA one so I guess this is a problem of configuration on our server.
Also @vmanthos propose to monitor tickets for that issue in case we missed something.

**Comment by piotrbak** [comment]
@CrochetFeve0251 Do we know why WP Rocket is trying to read that kind of paths? 🤔 
```
/shop-2/index-https.html/index-https.html_gzip
/shop-2/index-https.html/index-https.html
```

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 Do we know why WP Rocket is trying to read that kind of paths? thinking
> 
> ```
> /shop-2/index-https.html/index-https.html_gzip
> /shop-2/index-https.html/index-https.html
> ```

@piotrbak no idea but I guess somewhere later in the process we are not checking if `/index-https.html` is already present on the URL before adding it.

**Comment by piotrbak** [comment]
@Mai-Saad Is this one reproducible?
https://github.com/wp-media/wp-rocket/pull/5458#issuecomment-1326274396

If so, would be great to have exact steps there, as the following paths look very wrong:
```
/shop-2/index-https.html/index-https.html_gzip
/shop-2/index-https.html/index-https.html
```

**Comment by Mai-Saad** [comment]
@piotrbak I can see this warning on new labs, using the following steps:
Precondition:
- The shop page is set
- WP sitemap is used 

1- Fresh install to 3.12.3.3 zip
2- Visit the shop page and product page => pages are completed in the cache table
3- Edit the cached product => /shop/index-https.html URLs added
4- From the plugins page, update the zip of the fix using this [snippet](https://snippi.com/s/jnq6zyo) where the fix zip is 3.12.3.4
5- Wait till the index URLs are processed and removed from the cache table 
6- Check debug.log
```
[09-Dec-2022 10:08:39 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/new.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/new.rocketlabsqa.ovh/shop/index-https.html/index-https.html_gzip) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 96
[09-Dec-2022 10:08:39 UTC] PHP Warning:  is_readable(): open_basedir restriction in effect. File(/var/www/new.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/new.rocketlabsqa.ovh/shop/index-https.html/index-https.html) is not within the allowed path(s): (/var/www/:/usr/share/php/:/tmp/:/var/run/nginx-cache/:/dev/urandom:/dev/shm:/var/lib/php/sessions/) in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php on line 100
```

Notes:
- The error wasn't there when running the same scenario but with a custom sitemap containing a shop and product URLs
-  We may need to run the corn to process pending preload URLs

File: inc/ThirdParty/Plugins/Ecommerce/WooCommerceSubscriber.php [hunk 4]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace WP_Rocket\ThirdParty\Plugins\Ecommerce;
 
+use FakerPress\Provider\WP_Post;
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Ecommerce/WooCommerceSubscriber.php]: probably should not be there

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
can still see the error with the scenario mentioned [here](https://github.com/wp-media/wp-rocket/pull/5458#issuecomment-1344165350).

**Review by Mai-Saad — APPROVED** [review]
As per discussion with @piotrbak, we will leave the warning as is for now (it shouldn't occur for future updates and will only happen now if index.html wasn't processed yet before the update)

**Comment by remyperona** [comment]
Update needed to fix the conflicts

**Comment by piotrbak** [comment]
@CrochetFeve0251 There's update needed in this PR to fix conflicts

**Pull request merged**